pkg/runner/local_exec.go

```diff
@@ -32,29 +33,121 @@ var (
 type LocalExecutableRunnerCfg struct{}
 
+type healthcheckedProcess struct {
+	HealthcheckItem api.HealthcheckItem
+	Checker         func() bool
+	Fixer           func() error
+	Success         string
+	Failure         string
+}
+
+func newhealthcheckedProcess(ctx context.Context, name string, address string, cmd string, args ...string) *healthcheckedProcess {
+	return &healthcheckedProcess{
+		HealthcheckItem: api.HealthcheckItem{
+			Name: name,
+		},
+		Checker: tcpChecker(address),
+		Fixer:   commandStarter(ctx, cmd, args...),
+		Success: fmt.Sprintf("%s instance check: OK", name),
+		Failure: fmt.Sprintf("%s instance check: FAIL", name),
+	}
+}
+
+// tcpChecker returns a closure which can be used to check
+// when a tcp port is open. Returns true if the socket is dialable.
+// Otherwise, returns false. Use as healthcheckedProcess.Checker.
+func tcpChecker(address string) func() bool {
+	return func() bool {
+		_, err := net.Dial("tcp", address)
+		if err == nil {
+			return true
+		}
+		return false
+	}
+}
+
+func commandStarter(ctx context.Context, cmd string, args ...string) func() error {
+	return func() error {
+		cmd := exec.CommandContext(ctx, cmd, args...)
+		return cmd.Start()
+	}
+}
+
 func (r *LocalExecutableRunner) Healthcheck(fix bool, engine api.Engine, writer io.Writer) (*api.HealthcheckReport, error) {
 	r.lk.Lock()
 	defer r.lk.Unlock()
 
-	var redisCheck, outputsDirCheck api.HealthcheckItem
+	ctx, cancel := context.WithCancel(engine.Context())
+	r.closeFn = cancel
 
-	// Check if a local Redis instance is running. If not, try to start it.
-	_, err := net.Dial("tcp", "localhost:6379")
-	if err == nil {
-		msg := "local redis instance check: OK"
-		redisCheck = api.HealthcheckItem{Name: "local-redis", Status: api.HealthcheckStatusOK, Message: msg}
-	} else {
-		msg := fmt.Sprintf("local redis instance check: FAIL; %s", err)
-		redisCheck = api.HealthcheckItem{Name: "local-redis", Status: api.HealthcheckStatusFailed, Message: msg}
+	report := api.HealthcheckReport{
+		Checks: []api.HealthcheckItem{},
+		Fixes:  []api.HealthcheckItem{},
+	}
+
+	// Use this slice to add or remove additional checks.
+	localInfra := []*healthcheckedProcess{
+		newhealthcheckedProcess(ctx,
+			"local-redis",
+			"localhost:6379",
+			"redis-server",
+			"--save",
+			"\"\"",
+			"--appendonly",
+			"no"),
+		newhealthcheckedProcess(ctx,
+			"local-prometheus",
+			"localhost:9090",
+			"prometheus"),
+		newhealthcheckedProcess(ctx,
+			"local-pushgateway",
+			"localhost:9091",
+			"pushgateway"),
 	}
 
+	eg, ctx := errgroup.WithContext(ctx)
+
+	for _, li := range localInfra {
+		hcp := *li
+		eg.Go(func() error {
+			// Checker succeeds, already working.
+			if hcp.Checker() {
+				hcp.HealthcheckItem.Status = api.HealthcheckStatusOK
+				hcp.HealthcheckItem.Message = hcp.Success
+				report.Checks = append(report.Checks, hcp.HealthcheckItem)
+				return nil
+			}
+			// Checker failed, try to fix.
+			err := hcp.Fixer()
+			if err == nil {
```

Comment:
Nitpick, but probably a good idea to switch `err == nil` to `err != nil` - make it easier to read.